### PR TITLE
[Misc][ViT Cuda Graphs] Enable Piecewise CUDA Graphs for Qwen3-VL and Qwen2.5-VL ViT to Improve Performance

### PR DESCRIPTION
**Pull request by HirokenOvo**
## Purpose

Based on the torch.compile mechanism for generic nn.Module from [PR #23207](https://github.com/vllm-project/vllm/pull/23207) and [PR #27741](https://github.com/vllm-project/vllm/pull/27741), this PR implements piecewise CUDA graph support for the Vision Transformer (ViT) encoder in Qwen2.5-VL and Qwen3-VL models. The primary goal is to **eliminate kernel launch bubbles** for operators other than attention, thereby reducing overhead and improving inference performance.

This optimization is particularly effective for scenarios with low concurrency where the number of images is insufficient to fully saturate the ViT's computational capacity. By reducing kernel launch overhead, it helps to lower the TTFT.

## Key Features

1. **Supported Models**: Enabled for Qwen2.5-VL and Qwen3-VL series. Modified `Qwen2_5_VisionTransformer` and `Qwen3_VisionTransformer` to be graph-friendly, including handling of persistent buffers for hidden states and RoPE.
2. **Parallelism Modes**: Supports both ViT TP mode and DP mode for ViT CUDA graphs.

### Enable ViT CUDA Graph

To enable this feature, you need to set `compile_mm_encoder` to `True` via the `--compilation-config` argument.

```bash
--compilation-config '{"compile_mm_encoder": true}'
```

### Configure Capture Sizes

You can specify the image patch counts for which to capture ViT CUDA graphs using `mm_encoder_cudagraph_capture_sizes`.

```bash
--compilation-config '{"compile_mm_encoder": true, "mm_encoder_cudagraph_capture_sizes": [512, 1024]}'
```

If not specified, vLLM will automatically select a set of sizes based on the model's encoder budget.

Alternatively, you can specify `max_mm_encoder_cudagraph_capture_size` to generate a default list of capture sizes up to the given value:

```bash
--compilation-config '{"compile_mm_encoder": true, "max_mm_encoder_cudagraph_capture_size": 2048}'
```

## Limitations & Notes

- **Image Only**: This feature currently only supports **image inference**. Video inference is **not supported** yet.
- **`torch.compile` Issues**: `torch.compile` consumes a significant amount of GPU memory for compiling the ViT layer and may introduce negative optimization, leading to a slight increase in latency for individual operators. However, this issue is **orthogonal** to this PR, which focuses on enabling CUDA graph capture to reduce launch overhead.

## Performance

The performance was benchmarked under the following configuration:

- **Hardware**: H800
- **Model**: Qwen3-VL 32B
- **Parallelism**: TP=4, ViT DP=4
- **Input**: The total number of image patches processed per rank is calculated as: `(20 images / 4 DP ranks) * 128 tokens/image * 4 merge_size = 2560`.

without ViT cudagraph:
image

with ViT cudagraph:
image
image

The remaining bubbles are due to the GEMM being too small to fully hide the kernel launch overhead of the attention operator. Future work could involve implementing a full CUDA graph with FlashAttention-3 for the ViT to completely eliminate this overhead.

---

## Test Plan

Run the following test to verify the consistency between ViT CUDA graph execution and eager mode:

```bash
pytest tests/compile/piecewise/test_qwenvl_vit_cudagraph.py
```

## Test Result

6 passed, confirming the correctness of the piecewise CUDA graph implementation.

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
  </details>

### CODE REVIEW

File: vllm/compilation/backends.py [hunk 1]
```diff
@@ -48,6 +48,45 @@
...
+    is the first one in a sequence of graphs (e.g., a sequence of blocks).
+    """
+    global _is_first_graph_in_vit_sequence
+    original_value = _is_first_graph_in_vit_sequence
+    _is_first_graph_in_vit_sequence = is_first
+    try:
+        yield
+    finally:
+        _is_first_graph_in_vit_sequence = original_value
+
```
[the remark this answers is not in the document]
Comment (Lucaskabela) [reply vllm/compilation/backends.py]: I want to second this comment - I have had issues reasoning about and seen bugs introduced from these global variables from context managers.  If we can explore another way of passing this information, I think it would be much better for the code

File: vllm/config/vllm.py [hunk 2]
```diff
@@ -1275,6 +1283,117 @@ def _set_compile_ranges(self):
...
+            If these conditions are not met, the list of capture sizes will be empty,
+            effectively disabling ViT CUDA graphs.
+
+        2.  If the user has explicitly provided `vit_cudagraph_capture_sizes` in the
+            compilation config, those sizes are used. The list is de-duplicated
+            and sorted in ascending order.
+
+        3.  If no sizes are provided by the user, a default list of sizes is
+            generated up to a maximum of 5120. The default sizes are:
+            [512, 1024, 1536] + list(range(2048, 2048, 128)) + list(
```
Comment (Lucaskabela) [inline vllm/config/vllm.py]: Let's add a comment explaining why these are the default ranges we are using (i.e image sizes are usually one of these or something to that effect)

File: vllm/config/vllm.py [hunk 3]
```diff
@@ -1275,6 +1283,117 @@ def _set_compile_ranges(self):
...
+            - CUDA graph mode is enabled.
+            - The multimodal encoder compilation is enabled.
+            If these conditions are not met, the list of capture sizes will be empty,
+            effectively disabling ViT CUDA graphs.
+
+        2.  If the user has explicitly provided `vit_cudagraph_capture_sizes` in the
+            compilation config, those sizes are used. The list is de-duplicated
+            and sorted in ascending order.
+
+        3.  If no sizes are provided by the user, a default list of sizes is
```
Comment (Lucaskabela) [inline vllm/config/vllm.py:1363]: There are a few details that don't make sense to me in this comment; namely:

1) Maximum if 5120 - however, the `list(range(4096, 8192 + 1, 256))` will have a max of 8192
2) `list(range(2048, 2048, 128))`= [] since start and end are empty

File: vllm/compilation/backends.py [hunk 4]
```diff
@@ -48,6 +48,45 @@
 
 logger = init_logger(__name__)
 
+# A global flag to indicate if the current graph being compiled
```
Comment (Lucaskabela) [inline vllm/compilation/backends.py]: We are also using this compile_mm_encoder logic for audio encoders such as whisper - can we keep the language for naming agnostic of the modality?

**Review by Lucaskabela — CHANGES_REQUESTED** [review]

File: vllm/config/vllm.py (hunk 2 above)
Comment (HirokenOvo) [reply vllm/config/vllm.py]: > Let's add a comment explaining why these are the default ranges we are using (i.e image sizes are usually one of these or something to that effect)

@Lucaskabela Actually, in our internal usage, we specify `vit_cudagraph_capture_sizes` based on specific workload requirements. For the general community default, we designed it with the following considerations:

1. Patch Variance: Unlike the LLM part (which might use a step size of 8), image patch counts vary significantly.
2. Small Inputs: For scenarios with very few images, the GEMM computation is too small to effectively hide kernel launch overheads. Therefore, padding to a larger starting size is acceptable.
3. Step Size: Increasing the stride doesn't significantly add to the computation time for the ViT part. However, a larger stride reduces the number of graphs to be captured, which saves VRAM and reduces startup time.

We are not certain if this default fits all general community scenarios, so we are open to discussion if there are better configurations.

File: vllm/config/vllm.py (hunk 3 above)
Comment (HirokenOvo) [reply vllm/config/vllm.py:1363]: Sry. I changed the logic but missed updating the docstring to match. I've fixed the comments in the latest commit.

File: vllm/compilation/backends.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HirokenOvo) [reply vllm/compilation/backends.py]: > I want to second this comment - I have had issues reasoning about and seen bugs introduced from these global variables from context managers. If we can explore another way of passing this information, I think it would be much better for the code


@Lucaskabela The original reason for using global variables was that ViT `torch.compile` is applied piecewise to individual sub-modules, rather than the top-level module. This means the compilation backend only sees one sub-module's graph at a time, not the full ViT graph, rendering the existing first/last graph detection logic (used for LLMs) incorrect.

In the latest commit, I have refactored this to pass the information via the forward context, eliminating the need for these global variables. cc @wangxingran222

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -4925,7 +5059,50 @@ def _dummy_pooler_run(
...
+                set_forward_context(
+                    None,
+                    self.vllm_config,
+                    num_tokens=capture_size,
+                    cudagraph_runtime_mode=cudagraph_mode,
+                    batch_descriptor=batch_descriptor,
+                ),
+            ):
+                self.model.embed_multimodal(**dummy_mm_inputs)
+        self.vllm_config.is_in_compile_or_vit_cuda_graph_capture = False
```
[the remark this answers is not in the document]
Comment (HirokenOvo) [reply vllm/v1/worker/gpu_model_runner.py]: This method is executed only during `_dummy_mm_encoder_run` or `profile_run`. If an exception occurs here, the worker initialization constitutes a fatal error and the process will terminate. Therefore, resetting the flag is not strictly necessary as there will be no subsequent execution flow.

File: vllm/v1/worker/gpu_model_runner.py [hunk 6]
```diff
@@ -4402,6 +4499,43 @@ def rand_inputs_embeds() -> torch.Tensor:
             yield
             inputs_embeds.fill_(0)
 
+    def _get_dummy_h_w_patches(self, patches: int):
```
Comment (DarkLight1337) [inline vllm/v1/worker/gpu_model_runner.py]: IMO this should be defined by dummy data builder for each model, not inside model runner

**Review by DarkLight1337 — COMMENTED** [review]
Overall I'm not a fan of how much complexity this adds to both the model implementation and model runner. We should try to integrate this using similar principles as model runner v2.

File: vllm/v1/cudagraph_dispatcher.py [hunk 7]
```diff
@@ -221,6 +256,7 @@ def dispatch(
         has_lora: bool = False,
         disable_full: bool = False,
         num_active_loras: int = 0,
+        is_mm_encoder: bool = False,
```
Comment (LucasWilkinson) [inline vllm/v1/cudagraph_dispatcher.py]: Thanks for the contribution, the performance gains are impressive!

i agree with @DarkLight1337, the complexity it too high right now, in particular inside the model runner and `CudagraphDispatcher`, on that note

why not just instantiate another `cudagraph_dispatcher` in the gpu_model_runner

```
        # Cudagraph dispatcher for runtime cudagraph dispatching.
        self.cudagraph_dispatcher = CudagraphDispatcher(self.vllm_config)

        if self.supports_mm_inputs:
              self.mm_cudagraph_dispatcher = CudagraphDispatcher(self.vllm_config)
              ...
```

and initialize these with different keys. Then these flags are not required. We can `initialize_cudagraph_keys` to accept `capture_sizes` instead of fetching it directly from `self.compilation_config.cudagraph_capture_sizes` to assist with this. We do this for eagle (have a separate `cudagraph_dispatcher` instance)

Ideally we'd keep the  `CudagraphDispatcher` as simple as possible.

**Review by LucasWilkinson — CHANGES_REQUESTED** [review]

File: vllm/v1/cudagraph_dispatcher.py (hunk 7 above)
Comment (LucasWilkinson) [reply vllm/v1/cudagraph_dispatcher.py]: see: https://github.com/vllm-project/vllm/blob/61e632aea15f76fd1c46354b00f9cac62cd28c4e/vllm/v1/spec_decode/eagle.py#L107-L111
https://github.com/vllm-project/vllm/blob/61e632aea15f76fd1c46354b00f9cac62cd28c4e/vllm/v1/spec_decode/eagle.py#L272-L287

File: vllm/model_executor/models/qwen3_vl.py [hunk 8]
```diff
@@ -1405,46 +1533,70 @@ def _parse_and_validate_video_input(
...
                 self.visual.dtype
             )
-            if self.use_data_parallel:
-                grid_thw_list = grid_thw.tolist()
-                return run_dp_sharded_mrope_vision_model(
-                    self.visual, pixel_values_videos, grid_thw_list, rope_type="rope_3d"
-                )
-            else:
-                video_embeds = self.visual(pixel_values_videos, grid_thw=grid_thw)
+            with set_current_vllm_config(self.vllm_config):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen3_vl.py]: Is `set_current_vllm_config` really necessary? From my understanding it is already set as part of `Worker.load_model`

File: vllm/v1/cudagraph_dispatcher.py (hunk 7 above)
Comment (HirokenOvo) [reply vllm/v1/cudagraph_dispatcher.py]: Thanks for your suggestion. Following your advice has indeed significantly reduced the modifications required in `CudagraphDispatcher`. Regarding `gpu_model_runner`, I have encapsulated the relevant logic into `MMEncoderCudagraphManager` to minimize changes to the runner itself.

File: vllm/model_executor/models/qwen3_vl.py (hunk 8 above)
Comment (wangxingran222) [reply vllm/model_executor/models/qwen3_vl.py]: > Is `set_current_vllm_config` really necessary? From my understanding it is already set as part of `Worker.load_model`

@DarkLight1337 Yes, `set_current_vllm_config` isn't necessary anymore.

Previously in ViT DP mode, `vision.py` relied on `get_current_vllm_config` to fetch the config for `set_forward_context`. Since the new implementation allows `vision.py` to retrieve the `vllm_config` directly from the passed `MMEncoderCudagraphManager`, the global getter/setter logic is no longer needed.

File: vllm/v1/cudagraph_dispatcher.py [hunk 9]
```diff
@@ -29,9 +29,20 @@ class CudagraphDispatcher:
     runnable without cudagraph (if the mode does not match or mode is NONE).
     """
 
-    def __init__(self, vllm_config: VllmConfig):
+    def __init__(self, vllm_config: VllmConfig, is_mm_encoder: bool = False):
```
Comment (LucasWilkinson) [inline vllm/v1/cudagraph_dispatcher.py]: thanks for cleaning this up! lets just pass `max_capture_size` and `capture_sizes` so we can avoid this and make more extensible

Comment (LucasWilkinson) [reply vllm/v1/cudagraph_dispatcher.py]: (will be helpful if we have separate sizes for drafters too)

Comment (LucasWilkinson) [reply vllm/v1/cudagraph_dispatcher.py]: we should pass it into `initialize_cudagraph_keys` instead of `__init__` to so it happens after `adjust_cudagraph_sizes_for_spec_decode`

Comment (HirokenOvo) [reply vllm/v1/cudagraph_dispatcher.py]: Thanks for the suggestions! I have refactored `CudagraphDispatcher` as requested.

File: vllm/forward_context.py [hunk 10]
```diff
@@ -250,6 +250,10 @@ class ForwardContext:
     all_moe_layers: list[str] | None = None
     moe_layer_index: int = 0
 
+    # mm_encoder Multi-Modal Encoder flags used by backend compiler
```
Comment (DarkLight1337) [inline vllm/forward_context.py:253]: Am I missing something? Where are these flags being read?

Comment (wangxingran222) [reply vllm/forward_context.py:253]: > Am I missing something? Where are these flags being read?

@DarkLight1337 Sorry about the confusing comment. These flags are actually consumed by the Dynamo post-processing logic in `vllm/compilation/backends.py`.

Since ViT uses `torch.compile` on submodules (unlike the top-level approach in LLMs), we end up with multiple Dynamo graphs. The backend incorrectly treats each submodule as a standalone sequence, leading to wrong `gc_disable` and `weak_ref_output` settings. This causes intermediate tensors to be prematurely garbage collected during capture.

To fix this, we introduced these flags and manually call `set_is_first_graph_in_mm_encoder_sequence` and `set_is_last_graph_in_mm_encoder_sequence` inside the ViT `forward` method. This explicitly hints the correct global sequence boundaries to the backend wrapper.

**Comment by LucasWilkinson** [comment]
To avoid torch.compile we could maybe try going straight to FULL-cudagraphs? or is there something preventing that?

**Comment by wangshangsam** [comment]
> To avoid torch.compile we could maybe try going straight to FULL-cudagraphs? or is there something preventing that?

For the [MLPerf v6.0 Qwen3-VL](https://github.com/mlcommons/inference/tree/master/multimodal/qwen3-vl) submission from NVIDIA, our team actually developed full cudagraph (and piece-wise graph and torch.compile too, though slightly differently from this PR). We are happy to upstream this feature.

cc @b-mu @maxyanghu who developed it, and @ywang96 who's helping to upstream some other coming-from-MLPerf features currently.
